slapp/transforms/video_utils.py

```diff
@@ -46,32 +51,130 @@ def downsample_h5_video(
+
+    Parameters
+    ----------
+    video_paths : List[str]
+        A list of paths (str) for videos that should be concatenated together
+        Order of paths matters!
+    output_path : str
+        The desired output path for the concatenated video
+
+    Returns
+    -------
+    str
+        Path of concatenated video
     """
-    Function to transform 2p gray scale video into a webm
-    video using imageio_ffmpeg.
-    Args:
-        video: Video to be transformed with shape (time, row, col)
-        output_path: Output path for the transformed video
-        fps: desired fps of the output video
 
-    Returns:
+    with tempfile.NamedTemporaryFile(suffix=".txt") as concat_list_file:
+        with open(concat_list_file.name, 'w') as fp:
+            for path in video_paths:
+                fp.write(f"file '{path}'\n")
 
+        # See: https://trac.ffmpeg.org/wiki/Concatenate
+        ffmpeg_concat_cmd = [mpg.get_ffmpeg_exe(), '-y', '-f', 'concat',
+                             '-safe', '0', '-i', concat_list_file.name,
+                             '-c', 'copy', output_path]
+        subprocess.run(ffmpeg_concat_cmd)
+
+    return output_path
+
+
+def encode_video(video: np.ndarray, output_path: str,
+                 fps: float, bitrate: str = "0", crf: int = 20) -> str:
+    """Encode a video with vp9 codec via imageio-ffmpeg
+
+    Parameters
+    ----------
+    video : np.ndarray
+        Video to be encoded
+    output_path : str
+        Desired output path for encoded video
+    fps : float
+        Desired frame rate for encoded video
+    pix_fmt_out : str
+        Desired pixel format for output encoded video, by default "yuv420p"
+    bitrate : str, optional
+        Desired bitrate of output, by default "0". The default *MUST*
+        be zero in order to encode in constant quality mode. Other values
+        will result in constrained quality mode.
+    crf : int, optional
+        Desired perceptual quality of output, by default 20. Value can
+        be from 0 - 63. Lower values mean better quality (but bigger video
+        sizes).
+
+    Returns
+    -------
+    str
+        Output path of the encoded video
     """
 
-    # ffmpeg expects the video shape in width, height not row, col
-    # have to reverse shape when inputting
-    # gray8 is uint8 format
+    # ffmpeg expects video shape in terms of: (width, height)
+    video_shape = (video[0].shape[1], video[0].shape[0])
 
     writer = mpg.write_frames(output_path,
-                              (video[0].shape[1], video[0].shape[0]),
+                              video_shape,
                               pix_fmt_in="gray8",
                               pix_fmt_out="yuv420p",
                               codec="libvpx-vp9",
                               fps=fps,
-                              bitrate=bitrate)
-    writer.send(None)
+                              bitrate=bitrate,
+                              output_params=["-crf", str(crf)])
```

Comment:
Note: `crf` has to be set via `output_params` because `imageio-ffmpeg` does the wrong thing if you try to set `quality`:

https://github.com/imageio/imageio-ffmpeg/blob/master/imageio_ffmpeg/_io.py#L385-396